File: fsharp/isogram/Isogram.fs
module Isogram

open System

let private sanitize (str: string) =
    str.ToUpper ()
    |> Seq.filter Char.IsLetter
    |> Seq.toList

let isIsogram (str: string): bool = 
    let input = sanitize str

    (* simple solution:
    input.Length = (set input).Count
    *)

    (* more fun solution *)
    let rec helper chars seen =
        match chars with
        | [] -> true
        | c :: _ when Set.contains c seen -> false
        | c :: cs -> helper cs (Set.add c seen)

    helper input Set.empty

File: dart/isogram/lib/isogram.dart
final ALPHA = RegExp(r'\p{Alphabetic}', caseSensitive: false, unicode: true);

extension on String {
  bool isLetter() => ALPHA.hasMatch(this);
}

class Isogram {
  bool isIsogram(String word) {
    var seen = Set<String>();
    for (var char in word.toLowerCase().split('')) {
      if (char.isLetter()) {
        if (seen.contains(char)) return false;
        seen.add(char);
      }
    }
    return true;
  }
}
